README.md
# lein-sysutils

Query system information via Apache commons SystemUtils


[![Build Status](https://travis-ci.org/gonewest818/lein-sysutils.svg?branch=master)](https://travis-ci.org/gonewest818/lein-sysutils)
[![Code Coverage](https://codecov.io/gh/gonewest818/lein-sysutils/branch/master/graph/badge.svg)](https://codecov.io/gh/gonewest818/lein-sysutils)
[![Dependencies Status](https://jarkeeper.com/gonewest818/lein-sysutils/status.svg)](https://jarkeeper.com/gonewest818/lein-sysutils)
[![Clojars Project](https://img.shields.io/clojars/v/lein-sysutils.svg)](https://clojars.org/lein-sysutils)

## Configuration:

    :plugins [[lein-sysutils "0.2.0"]]

## Usage:

    lein sysutils [:edn] :key1 [:key2 :key3 ...]

Sysutils provides a command line interface to print and inspect
metadata about the current system config exposed through the Apache
Commons SystemUtils API. The fields are presented as a hashmap, and
can be queried by passing in one or more "keywordized" field names.

The output is available in two forms. By default keywords and values
are printed as pairs, separated with whitespace, one pair per line. If
the optional `:edn` key is set, then the requested keys are printed as EDN.
Depending on your situation you may find one format more convenient than
the other.

To make life easier for devops, the plugin provides an additional
query that is not directly available in the SystemUtils library. The
`:java-version-simple` key is mapped to a string that indicates which
"release" of Java is currently in use. Because of inconsistent
naming across vendors and releases, here the string with be normalized
to "1.1", "1.2", ... "1.8", or for newer releases "9" and "10".

## Examples:

### Print the "simple" Java version number

    lein sysutils :java-version-simple

### Print the Java VM name

    lein sysutils :java-vm-name

### Output multiple values as parseable EDN

    lein sysutils :edn :java-vm-name :java-home

## Available Keys:

    :awt-toolkit
    :file-encoding
    :file-separator
    :is-java-1-1
    :is-java-1-2
    :is-java-1-3
    :is-java-1-4
    :is-java-1-5
    :is-java-1-6
    :is-java-1-7
    :is-java-1-8
    :is-java-1-9
    :is-java-9
    :is-os-400
    :is-os-aix
    :is-os-free-bsd
    :is-os-hp-ux
    :is-os-irix
    :is-os-linux
    :is-os-mac
    :is-os-mac-osx
    :is-os-mac-osx-cheetah
    :is-os-mac-osx-el-capitan
    :is-os-mac-osx-jaguar
    :is-os-mac-osx-leopard
    :is-os-mac-osx-lion
    :is-os-mac-osx-mavericks
    :is-os-mac-osx-mountain-lion
    :is-os-mac-osx-panther
    :is-os-mac-osx-puma
    :is-os-mac-osx-snow-leopard
    :is-os-mac-osx-tiger
    :is-os-mac-osx-yosemite
    :is-os-net-bsd
    :is-os-open-bsd
    :is-os-os2
    :is-os-solaris
    :is-os-sun-os
    :is-os-unix
    :is-os-windows
    :is-os-windows-10
    :is-os-windows-2000
    :is-os-windows-2003
    :is-os-windows-2008
    :is-os-windows-2012
    :is-os-windows-7
    :is-os-windows-8
    :is-os-windows-95
    :is-os-windows-98
    :is-os-windows-me
    :is-os-windows-nt
    :is-os-windows-vista
    :is-os-windows-xp
    :is-os-zos
    :java-awt-fonts
    :java-awt-graphicsenv
    :java-awt-headless
    :java-awt-printerjob
    :java-class-path
    :java-class-version
    :java-compiler
    :java-endorsed-dirs
    :java-ext-dirs
    :java-home
    :java-io-tmpdir
    :java-library-path
    :java-runtime-name
    :java-runtime-version
    :java-specification-name
    :java-specification-vendor
    :java-specification-version
    :java-util-prefs-preferences-factory
    :java-vendor
    :java-vendor-url
    :java-version
    :java-version-simple
    :java-vm-info
    :java-vm-name
    :java-vm-specification-name
    :java-vm-specification-vendor
    :java-vm-specification-version
    :java-vm-vendor
    :java-vm-version
    :line-separator
    :os-arch
    :os-name
    :os-version
    :path-separator
    :user-country
    :user-dir
    :user-home
    :user-language
    :user-name
    :user-timezone

For more information on the available keys and what they mean, see the
Apache Documentation:

https://commons.apache.org/proper/commons-lang/javadocs/api-3.5/org/apache/commons/lang3/SystemUtils.html

## License

Copyright © 2018 Neil Okamoto

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## Unreleased
### Changed
- [#1] Ensure CI jobs running in forks don't attempt to deploy to our Clojars account.

## 0.2.0
### Changed
- First release


test/leiningen/sysutils_test.clj
(ns leiningen.sysutils-test
  (:require [clojure.test :refer :all]
            [leiningen.sysutils :refer :all]))

(deftest assoc-java-version-simple-test
  (let [m (assoc-java-version-simple {})]
    (is (= "unknown" (:java-version-simple m))) "unknown java version")
  (let [m (assoc-java-version-simple {:is-java-1-1 true})]
    (is (= "1.1" (:java-version-simple m)) "known java version")))

(deftest add-java-versions-test
  (let [m (add-java-versions {:java-specification-version "11"})]
    (is (false? (:is-java-10 m)))
    (is (true? (:is-java-11 m)))
    (is (false? (:is-java-12 m))))
  (let [m (add-java-versions {:is-java-10 false
                              :is-java-12 true})]
    (is (false? (:is-java-10 m)))
    (is (true? (:is-java-12 m)))))

(deftest extend-is-java-x-test
  (let [m (sysutils-map)]
    (is (= (:java-specification-version m)
           (:java-version-simple m)))))

(deftest keywordize-test
  (is (= :simple      (keywordize "SIMPLE")))
  (is (= :two-words   (keywordize "TWO_WORDS")))
  (is (= :digits-1234 (keywordize "DIGITS_1234")))
  (is (= :dots.ok     (keywordize "DOTS.OK"))))

(deftest format-output-test
  (let [m {:a true :b "quoted string"}]
    (is (= "{:a true, :b \"quoted string\"}\n"
           (with-out-str (format-output m :edn))))
    (is (= ":a true\n:b quoted string\n"
           (with-out-str (format-output m nil))))))

(deftest help-test
  (is (= "Query system information via Apache commons SystemUtils"
         (first (clojure.string/split-lines (help))))))

(deftest sysutils-map-test
  (with-redefs [sysutils-public-fields (constantly '("FOO" "BAR" "MY_FIELD"))
                sysutils-lookup {"FOO" "foo" "BAR" "bar" "MY_FIELD" true}]
    (let [m (sysutils-map)]
      (is (= "foo" (:foo m)))
      (is (= "bar" (:bar m)))
      (is (= true (:my-field m)))
      (is (= "unknown" (:java-version-simple m))))))

(deftest sysutils-lookup-test
  (is (= (System/getenv "USER") (sysutils-lookup "USER_NAME")))
  (is (= false (sysutils-lookup "IS_JAVA_1_1"))))

(deftest main-test
  (with-redefs [sysutils-public-fields (constantly '("FOO" "BAR" "MY_FIELD"))
                sysutils-lookup {"FOO" "foo" "BAR" "bar" "MY_FIELD" true}]
    (is (= ":foo foo\n"
           (with-out-str (sysutils {} ":foo"))))
    (is (= "{:foo \"foo\"}\n"
           (with-out-str (sysutils {} ":edn" ":foo"))))
    ;; currently can't guarantee order of results
    (is (#{":foo foo\n:bar bar\n" ":bar bar\n:foo foo\n"}
         (with-out-str (sysutils {} ":foo" ":bar"))))
    ;; currently can't guarantee order of results
    (is (#{"{:foo \"foo\", :bar \"bar\"}\n" "{:bar \"bar\", :foo \"foo\"}\n"}
         (with-out-str (sysutils {} ":edn" ":foo" ":bar"))))))


project.clj
(defproject lein-sysutils "0.2.1-SNAPSHOT"
  :description "Query Java system parameters in Leiningen"
  :url "https://github.com/gonewest818/lein-sysutils"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :eval-in-leiningen true

  :aliases {"bump-version" ["change" "version" "leiningen.release/bump-version"]}

  :release-tasks [["vcs" "assert-committed"]
                  ["bump-version" "release"]
                  ["vcs" "commit" "Release %s"]
                  ["vcs" "tag" "v" "--no-sign"]
                  ["bump-version"]
                  ["vcs" "commit" "Begin %s"]]

  :deploy-repositories [["clojars" {:url "https://clojars.org/repo"
                                    :username :env/clojars_username
                                    :password :env/clojars_password
                                    :sign-releases false}]]

  :profiles {;; Versions matrix
             :1.7 {:dependencies [[org.clojure/clojure "1.7.0"]]}
             :1.8 {:dependencies [[org.clojure/clojure "1.8.0"]]}
             :1.9 {:dependencies [[org.clojure/clojure "1.9.0"]]}
             :master {:repositories [["snapshots"
                                      "https://oss.sonatype.org/content/repositories/snapshots"]]
                      :dependencies [[org.clojure/clojure "1.10.0-master-SNAPSHOT"]]}

             ;; CI
             :cloverage {:plugins [[lein-cloverage "1.0.10"]]}
             :cljfmt {:plugins [[lein-cljfmt "0.5.7"]]}
             :eastwood {:plugins [[jonase/eastwood "0.2.5"]]}})


src/leiningen/sysutils.clj
(ns leiningen.sysutils
  (:require [clojure.reflect :refer [reflect]])
  (:import (org.apache.commons.lang3 SystemUtils)))

(defn assoc-java-version-simple
  "Calculate what is colloquially thought of as the \"version\" of the
  JDK for purposes of integration. These are strings like \"1.6\",
  \"1.7\", \"1.8\" and \"9\".  Associate this value in the provided
  map with the key `:java-version-simple`."
  [m]
  (let [sv (cond
             (:is-java-1-1 m) "1.1"
             (:is-java-1-2 m) "1.2"
             (:is-java-1-3 m) "1.3"
             (:is-java-1-4 m) "1.4"
             (:is-java-1-5 m) "1.5"
             (:is-java-1-6 m) "1.6"
             (:is-java-1-7 m) "1.7"
             (:is-java-1-8 m) "1.8"
             (:is-java-9   m)   "9" ; new naming per JEP 223
             (:is-java-10  m)  "10" ; anticipating commons-lang >= 3.7
             :else            "unknown")]
    (assoc m :java-version-simple sv)))

(defn sysutils-public-fields
  "Find the names of the public fields in SystemUtils via reflection."
  []
  (->> (reflect SystemUtils)
       :members
       (filter (partial instance? clojure.reflect.Field))
       (filter (comp :public :flags))
       (map :name)))

(defn sysutils-lookup
  "Lookup the value of a SystemUtils field by name."
  [n]
  (eval (symbol (str "org.apache.commons.lang3.SystemUtils/" n))))

(defn keywordize
  "Convert a Java field name to an idiomatic Clojure keyword."
  [n]
  (-> (clojure.string/replace n #"_" "-") clojure.string/lower-case keyword))

(defn sysutils-map
  "Return the SystemUtils data as a map. The keys are the field names
  translated to idiomatic Clojure keywords. For example the resulting
  map will have the key `:is-java-9` associated with the value of
  `SystemUtils/IS_JAVA_9`, etc.  Constructing the map dynamically
  allows the available values to adjust to the implementation of the
  commons-lang3 library we are built upon."
  []
  (let [names  (sysutils-public-fields)
        keys   (map keywordize names)
        values (map sysutils-lookup names)]
    (assoc-java-version-simple (zipmap keys values))))

(defn format-output
  "Print the map m. If the edn parameter is false then print exactly one
  key and value pair per line. If the edn parameter is true then print
  the results as a map."
  [m edn]
  (if edn
    (println (pr-str m))
    (doseq [[k v] m]
      (println k v))))

(defn help []
  (with-out-str
    (println "Query system information via Apache commons SystemUtils")
    (println "
Usage: lein sysutils [:edn] :key1 [:key2 :key3 ...]

Sysutils provided a command line interface to print and inspect
metadata about the current system config exposed through the Apache
Commons SystemUtils API. The fields are presented as a hashmap, and
can be queried by passing in one or more \"keywordized\" field names.

The output is available in two forms. By default keywords and values
are printed as pairs, separated with whitespace, one pair per line. If
the optional `:edn` key is set, then the requested keys are printed as EDN.
Depending on your situation you may find one format more convenient than
the other.

To make life easier for devops, the plugin provides an additional
query that is not directly available in the SystemUtils library. The
`:java-version-simple` key is mapped to a string that indicates which
\"release\" of Java is currently in use. Because of inconsistent
naming across vendors and releases, here the string with be normalized
to \"1.1\", \"1.2\", ... \"1.8\", or for newer releases \"9\" and \"10\".

Examples:

* Print the \"simple\" Java version number
    lein sysutils :java-version-simple

* Print the Java VM name
    lein sysutils :java-vm-name

* Output multiple values as parseable EDN
    lein sysutils :edn :java-vm-name :java-home

Available Keys:
")
    (doseq [k (sort (keys (sysutils-map)))]
      (println (str "    " k)))
    (println "
For more information on the available keys and what they mean, see the
Apache Documentation:

https://commons.apache.org/proper/commons-lang/javadocs/api-3.5/org/apache/commons/lang3/SystemUtils.html")))

(defn ^:no-project-needed sysutils
  "Look up system information via the Apache commons-lang3 API."
  [project & args]
  (let [all     (set (->> args
                          (map clojure.edn/read-string)
                          (filter keyword?))) ; ignore non-keywords for now
        edn     (:edn all)
        queries (remove #{:edn} all)]
    (format-output (select-keys (sysutils-map) queries) edn)))
